src/garage/torch/algos/dqn.py

```diff
@@ -0,0 +1,268 @@
+                samples = self.replay_buffer.sample_transitions(
+                    self._buffer_batch_size)
+                samples['rewards'] *= self._reward_scale
+                qf_loss, y, q = tuple(v.cpu().numpy()
+                                      for v in self.optimize_qf(samples))
+
+                self._episode_qf_losses.append(qf_loss)
+                self._epoch_ys.append(y)
+                self._epoch_qs.append(q)
+
+        if itr % self._steps_per_epoch == 0:
+            logger.log('Training finished')
+
+            if (self.replay_buffer.n_transitions_stored >=
+                    self._min_buffer_size):
+                tabular.record('Epoch', epoch)
+                tabular.record('QFunction/AverageQFunctionLoss',
+                               np.mean(self._episode_qf_losses))
+                tabular.record('QFunction/AverageQ', np.mean(self._epoch_qs))
+                tabular.record('QFunction/MaxQ', np.max(self._epoch_qs))
+                tabular.record('QFunction/AverageAbsQ',
+                               np.mean(np.abs(self._epoch_qs)))
+                tabular.record('QFunction/AverageY', np.mean(self._epoch_ys))
+                tabular.record('QFunction/MaxY', np.max(self._epoch_ys))
+                tabular.record('QFunction/AverageAbsY',
+                               np.mean(np.abs(self._epoch_ys)))
+
+        if itr % self._target_update_freq == 0:
+            self._target_qf = copy.deepcopy(self._qf)
+
+    def optimize_qf(self, samples_data):
+        """Perform algorithm optimizing.
+
+        Args:
+            samples_data (dict): Processed batch data.
+
+        Returns:
+            qval_loss: Loss of Q-value predicted by the Q-network.
+            ys: y_s.
+            qval: Q-value predicted by the Q-network.
+
+        """
+        transitions = dict_np_to_torch(samples_data)
+
+        observations = transitions['observations']
+        rewards = transitions['rewards'].reshape(-1, 1)
+        actions = transitions['actions']
+        next_observations = transitions['next_observations']
+        terminals = transitions['terminals'].reshape(-1, 1)
+
+        next_inputs = next_observations
+        inputs = observations
+        with torch.no_grad():
+            # discrete, outputs Qs for all possible actions
+            target_qvals = self._target_qf(next_inputs)
+            best_qvals, _ = torch.max(target_qvals, 1)
+            best_qvals = best_qvals.unsqueeze(1)
+
+        rewards_clipped = rewards
+        if self._clip_reward is not None:
+            rewards_clipped = torch.clamp(rewards, -1 * self._clip_reward,
+                                          self._clip_reward)
+        y_target = (rewards_clipped +
+                    (1.0 - terminals) * self._discount * best_qvals)
+        y_target = y_target.squeeze(1)
+
+        # optimize qf
+        qvals = self._qf(inputs)
+        selected_qs = torch.sum(qvals * actions, axis=1)
```

Comment:
Ideally, can you break out your losses into separate functions, and write tests for them to ensure that they always maintain the same values given the same inputs. Those types of tests are the strongest for protecting against performance regressions.

Reply:
Tbh I'd rather keep the loss computation in the optimize function, it makes it much simpler to read. Other algos (VPG comes to mind) have multiple functions that compute various parts of the loss and I find that harder to follow. I agree that the loss should be tested though so I added a test to make sure the computed loss is what is expected.

Reply:
as long as the important part fits on ~1 screen this is ok with me. if it gets bigger, you should split it up.